src/analyses/ai_history.h

```diff
@@ -0,0 +1,229 @@
+/*******************************************************************\
+
+Module: Abstract Interpretation
+
+Author: Martin Brain, martin.brain@cs.ox.ac.uk
+
+\*******************************************************************/
+
+/// \file
+/// Abstract Interpretation history
+///
+/// This is an abstraction of the history of the program counter
+/// / the history of the execution of the program.
+/// It is used to track and control how the abstract interpreter
+/// explores the program and for context sensitive analyses it is
+/// also used as the key for the state map.
+///
+/// The abstract interpreter works with instances of ai_history_baset
+/// (and history_ptrt's when dealing with storage) but this is just
+/// an abstract interface much like ai_domain_baset, actual implementations
+/// derive from it.
+
+#ifndef CPROVER_ANALYSES_AI_HISTORY_H
+#define CPROVER_ANALYSES_AI_HISTORY_H
+
+#include <memory>
+
+#include <goto-programs/goto_model.h>
+
+#include <util/json.h>
+#include <util/xml.h>
+
+/// We make heavy use of reference counted pointers to history objects
+class ai_history_baset;
+
+class history_ptrt
+{
+public:
+  typedef std::shared_ptr<const ai_history_baset> ptrt;
+
+  history_ptrt(ptrt _p) : p(_p)
+  {
+  }
+
+  history_ptrt(const ai_history_baset *h) : p(h)
+  {
+  }
+
+  // Order and compare on the contents not the pointers
+  bool operator<(const history_ptrt &op) const;
+  bool operator==(const history_ptrt &op) const;
+
+  const ai_history_baset &operator*(void) const
+  {
+    return *p;
+  }
+
+private:
+  ptrt p;
+};
+
+/// This is the base interface; derive from this.
+class ai_history_baset
+{
+public:
+  typedef goto_programt::const_targett locationt;
+  typedef irep_idt function_namet;
+
+  ai_history_baset(const ai_history_baset &)
+  {
+  }
+
+  /// Create a new history starting from a given location
+  /// This is used to start the analysis from scratch
+  /// PRECONDITION(l.is_dereferenceable());
+  explicit ai_history_baset(locationt)
+  {
+  }
+
+  enum class step_statust
+  {
+    NEW_FORCE_CONTINUE,
+    NEW,
+    MERGED,
+    BLOCKED
+  };
+
+  typedef std::pair<step_statust, history_ptrt> step_returnt;
+  typedef std::set<history_ptrt> history_sett;
+
+  /// Step creates a new history by advancing the current one to location "to"
+  /// It is given the set of all other histories that have reached this point.
+
+  virtual step_returnt step(locationt to, history_sett &others) const = 0;
+
+  /// PRECONDITION(to.is_dereferenceable());
+  /// PRECONDITION(to \in goto_program.get_successors(current_location()) ||
+  ///              current_location()->is_function_call() ||
+  ///              current_location()->is_end_function());
+  ///
+  /// Step may do one of four things :
+  ///  1. Create a new history object in others and return a pointer to it.
+  ///     Always adds the history to the worklist.
+  ///     POSTCONDITION(IMPLIES(result.first ==step_statust::NEW_FORCE_CONTINUE,
```

Comment:
== spacing

Reply:
Good place to say why you'd do such a thing?

Reply:
Eagle eyes for spotting the trick I used to get this in 80 columns.  It's ugly but splitting it over two lines is worse IMHO.

Reply:
Reasons... it's a bit complicated.  This bit is still evolving a little as I'm working on other contexts.  One use case is the one given in ahistorical, where you don't want to not explore parts of the program because you're not enlarging the domains (for example, location insensitive or the fully flow insensitive).  You probably want to do this if you are creating the first history in a set.  If your history has other external effects (like running the binary in sync with the AI) then again you want to force continue.

Reply:
Please check doxygen whether these comments appear somewhere sensible.